cartographer_grpc/framework/server.cc

```diff
@@ -66,12 +76,40 @@ void Server::RunCompletionQueue(
   void* tag;
   while (completion_queue->Next(&tag, &ok)) {
     auto* rpc_event = static_cast<Rpc::RpcEvent*>(tag);
-    if (auto rpc = rpc_event->rpc.lock()) {
-      rpc->service()->HandleEvent(rpc_event->event, rpc.get(), ok);
-    } else {
-      LOG(WARNING) << "Ignoring stale event.";
+    rpc_event->ok = ok;
+    rpc_event->event_queue->Push(rpc_event);
+  }
+}
+
+void Server::ProcessRpcEvent(Rpc::RpcEvent* rpc_event) {
+  if (auto rpc = rpc_event->rpc.lock()) {
+    rpc->service()->HandleEvent(rpc_event->event, rpc.get(), rpc_event->ok);
+  } else {
+    LOG(WARNING) << "Ignoring stale event.";
+  }
+  delete rpc_event;
+}
+
+EventQueue* Server::SelectNextEventQueueRoundRobin() {
+  ++current_event_queue_id_;
+  current_event_queue_id_ =
+      current_event_queue_id_ % options_.num_event_threads;
+  return event_queue_threads_.at(current_event_queue_id_).event_queue();
+}
+
+void Server::RunEventQueue(EventQueue* event_queue) {
+  while (!shutting_down_) {
+    Rpc::RpcEvent* rpc_event = event_queue->PopWithTimeout(
+        cartographer::common::FromMilliseconds(100));
+    if (rpc_event) {
+      ProcessRpcEvent(rpc_event);
     }
-    delete rpc_event;
+  }
+
+  // Finish processing the rest of the items.
+  while (Rpc::RpcEvent* rpc_event = event_queue->PopWithTimeout(
+             cartographer::common::FromMilliseconds(100))) {
+    ProcessRpcEvent(rpc_event);
```

Comment:
This doesn't necessarily terminate. Can you just return and ignore the rest if the queue blocks for a certain time?

Reply:
As discussed.